<launch>
    <param name="/use_sim_time" value="true" />

    <arg name="model" default="burger" doc="model type [burger, waffle, waffle_pi]"/>

    <!-- Gazebo TurtleBot3 World -->
    <include file="$(find turtlebot3_gazebo)/launch/turtlebot3_world.launch"/>

    <!-- TurtleBot3 Bringup -->
    <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch"/>

    <!-- TF: map -> odom -->
    <node pkg="tf" type="static_transform_publisher" name="map_odom_broadcaster"
          args="0 0 0 0 0 0 map odom 100" />

    <!-- 静态地图服务器 -->
    <node name="map_server" pkg="map_server" type="map_server"
          args="$(find turtlebot3_navigation)/maps/map.yaml">
        <param name="frame_id" value="map"/>
    </node>

    <!-- 二层架构控制器 (A* + MPPI) - Nav2 风格 -->
    <node name="mppi_two_layer_controller" pkg="simple_mppi" type="mppi_two_layer.py" output="screen">
        <!-- PathHandler 参数 -->
        <param name="lookahead_dist" value="0.3"/>
        <param name="prune_distance" value="0.2"/>
        <param name="v_ref" value="0.15"/>

        <!-- Critics 权重 -->
        <param name="w_path_follow" value="50.0"/> 
        <param name="w_path_align" value="30.0"/>
        <param name="w_goal_dist" value="20.0"/>
        <param name="w_velocity" value="5.0"/>
        <param name="w_angular" value="1.0"/>
        <param name="w_obstacle" value="100.0"/>
        <param name="w_constraint" value="500.0"/>

        <!-- 障碍物参数 (APF 备用) -->
        <param name="obs_d0" value="0.4"/>
        <param name="obs_eta" value="1.0"/>
    </node>

    <!-- RViz -->
    <node name="rviz" pkg="rviz" type="rviz"
          args="-d $(find simple_mppi)/rviz/mppi_two_layer.rviz"/>
</launch>
